tensorboard/data/server/gcs/auth.rs

```diff
@@ -0,0 +1,300 @@
+                    creds_file, e
+                );
+                return Credentials::Anonymous;
+            }
+            Ok(f) => f,
+        };
+        match serde_json::from_reader(reader) {
+            Ok(refresh_token) => {
+                info!(
+                    "Using refresh token GCS credentials from {}",
+                    creds_file.display()
+                );
+                Credentials::RefreshToken(RefreshToken(refresh_token))
+            }
+            Err(e) => {
+                warn!(
+                    "Failed to read GCS credentials file {:?}; will use anonymous credentials: {}",
+                    creds_file, e
+                );
+                Credentials::Anonymous
+            }
+        }
+    }
+
+    /// Determines the file on disk from which credentials might be read, if any.
+    fn credentials_file() -> Option<PathBuf> {
+        if let Some(p) = std::env::var_os("GOOGLE_APPLICATION_CREDENTIALS") {
+            return Some(p.into());
+        }
+        if let Some(base_config_dir) = std::env::var_os("XDG_CONFIG_HOME")
+            .map(PathBuf::from)
+            .or_else(|| std::env::var_os("HOME").map(|p| PathBuf::from(p).join(".config")))
+        {
+            let mut path = base_config_dir;
+            path.extend(&["gcloud", "application_default_credentials.json"]);
+            if path.is_file() {
+                return Some(path);
+            }
+        };
+        None
+    }
+
+    /// Tests whether this credential is inherently anonymous. If this returns `true`, then
+    /// [`Self::fetch`] will always return `None`.
+    ///
+    /// This exists as an optimization so that a [`TokenStore`] doesn't need to check locks all the
+    /// time when the
+    fn anonymous(&self) -> bool {
+        matches!(self, Credentials::Anonymous)
+    }
+
+    /// Attempts to fetch a fresh access token with these credentials.
+    fn fetch(&self, http: &HttpClient) -> Option<BoundedToken> {
+        match self {
+            Credentials::Anonymous => None,
+            Credentials::RefreshToken(RefreshToken(creds)) => match creds.fetch(http) {
+                Ok(t) => Some(t),
+                Err(e) => {
+                    warn!("GCS authentication failed: {}", e);
+                    None
+                }
+            },
+        }
+    }
+}
+
+/// Persistent credentials in the form of an OAuth refresh token. Can be posted to an OAuth token
+/// endpoint to obtain a short-lived access token.
+#[derive(Deserialize)]
+struct RefreshTokenCreds {
+    client_id: String,
+    client_secret: String,
+    refresh_token: String,
+}
+impl Debug for RefreshTokenCreds {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("RefreshTokenCreds")
+            .field("client_id", &self.client_id)
+            .field("client_secret", &self.client_secret)
+            .field("refresh_token", &format_args!("<redacted>"))
```

Comment:
Just checking: if the `<redacted>` here is to cover the case when we accidentally save/log our refresh tokens somewhere, should we redact the client_secret too?

Reply:
…but, you know, on second thought, it costs nothing to hide the client
“secret”, so I might as well, just in case my understanding is missing
some pieces. Done.

Reply:
Yeah - it's not secret for well-known installed apps like `gcloud` or our uploader, since it's not possible to hide it, but I think it could be a real secret for a generic `GOOGLE_APPLICATION_CREDENTIALS` path.